### PR TITLE
Revise tactic info selection

### PR DESCRIPTION
**Pull request by Kha**
Lean's info tree contains the initial and final states of all tactics and tactic combinators executed in a tactic block, but there is some ambiguity as to which state we should show at which location. In particular, in contrast to Lean 3 we would like to show *all* intermediate states when hovering over a tactic that is run multiple times by a combinator such as `repeat`. I implemented a new heuristic in https://github.com/leanprover/lean4/commit/eb7bf2b272862940832ed63a5892a5c261fbe446 that forced the info view to show only the last such state *after* e.g. `repeat tac`, but this turned out to be less desirable inside combinator chains as in #1119.

This PR reverts the above commit, replacing it with a different but more general heuristic that also solves many other deficiencies: tactic info selection is now indentation-sensitive. Thus, while at the *end* of something like `next => repeat b` it is hopelessly ambiguous on what level the user wants to continue the proof, so we show the innermost infos, the repeated executions of `b` as before, if the user first inserts a line break, we select the info of `next`, `repeat`, or `b` depending on the indentation of the new line. Crucially, this also solves the issue of returning to the main goal after finishing a `have` sub-proof: as soon as you deindent back to the level of the `have` keyword, the main goal will be shown.

### CODE REVIEW

File: tests/lean/interactive/infoIssues.lean [hunk 1]
```diff
@@ -1,28 +1,30 @@
...
   match x with
   | 0 => rfl
-  | y+1 => repeat rw [Nat.zero_add] /- Good: We get the correct behavior here: Goals accomplished. -/
-       --^ Bad: get the `match` tactic initial state here.
-  -- Bad: we get "Goals accomplished" here
+                                 -- Bad? We still get all states except the one closed by implicit `rfl`
+                                 --v $/lean/plainGoal
+  | y+1 => repeat rw [Nat.zero_add]
+       --^ $/lean/plainGoal
+       -- Bad: get the `match` tactic initial state here.
```
Comment (Kha) [inline tests/lean/interactive/infoIssues.lean:28]: NB: this is an orthogonal issue in the info tree, not info selection

Comment (Kha) [reply tests/lean/interactive/infoIssues.lean:28]: I have an in-progress solution to #1119 that should be applicable here as well

File: src/Lean/Server/InfoUtils.lean [hunk 2]
```diff
@@ -25,23 +25,31 @@ def Lean.Syntax.getRange? (stx : Syntax) (originalOnly := false) : Option String
...
-        | some a => [a]
-        | _      => []
+    match ctx? with
+      | some ctx => p ctx i cs cs'
       | _        => []
   | _ => []
 
+/--
+  For every branch, find the deepest node in that branch matching `p`
+  with a surrounding context (the innermost one) and return all of them. -/
```
Comment (Vtec234) [inline src/Lean/Server/InfoUtils.lean:43]: I found this a little difficult to understand.
```suggestion
  For every branch of the `InfoTree`, find the deepest node in that branch for which `p` returns
  `some _`  and return the union of all such nodes. The visitor `p` is given a node together with
  its innermost surrounding `ContextInfo`. -/
```
